frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1525 @@
+					// next 2 eras because their vote weight will not be counted until the
+					// snapshot in active era + 1.
+					reward_pool_total_earnings: reward_pool.total_earnings,
+					unbonding_era: None,
+				},
+			);
+			bonded_pool.put();
+			Self::deposit_event(Event::<T>::Joined {
+				delegator: who,
+				pool: pool_account,
+				bonded: amount,
+			});
+
+			Ok(())
+		}
+
+		/// A bonded delegator can use this to claim their payout based on the rewards that the pool
+		/// has accumulated since their last claimed payout (OR since joining if this is there first
+		/// time claiming rewards).
+		///
+		/// Note that the payout will go to the delegator's account.
+		#[pallet::weight(T::WeightInfo::claim_payout())]
+		pub fn claim_payout(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let delegator = Delegators::<T>::get(&who).ok_or(Error::<T>::DelegatorNotFound)?;
+			let bonded_pool = BondedPool::<T>::get(&delegator.pool)
+				.defensive_ok_or_else(|| Error::<T>::PoolNotFound)?;
+
+			Self::do_reward_payout(who, delegator, &bonded_pool)?;
+
+			Ok(())
+		}
+
+		/// Unbond _all_ of the `target` delegators funds from the pool. Under certain conditions,
+		/// this call can be dispatched permissionlessly (i.e. by any account).
+		///
+		/// Conditions for a permissionless dispatch:
+		///
+		/// - The pool is blocked and the caller is either the root or state-toggler. This is
+		///   refereed to as a kick.
+		/// - The pool is destroying and the delegator is not the depositor.
+		/// - The pool is destroying, the delegator is the depositor and no other delegators are in
+		///   the pool.
+		///
+		/// Conditions for permissioned dispatch (i.e. the caller is also the target):
+		///
+		/// - The caller is not the depositor
+		/// - The caller is the depositor, the pool is destroying and not other delegators are in
+		///   the pool.
+		///
+		/// Note: If their are too many unlocking chunks to unbond with the pool account,
+		/// [`Self::withdraw_unbonded_pool`] can be called to try and minimize unlocking chunks.
+		#[pallet::weight(666)]
+		pub fn unbond_other(origin: OriginFor<T>, target: T::AccountId) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+			let delegator = Delegators::<T>::get(&target).ok_or(Error::<T>::DelegatorNotFound)?;
+			let mut bonded_pool = BondedPool::<T>::get(&delegator.pool)
+				.defensive_ok_or_else(|| Error::<T>::PoolNotFound)?;
+			bonded_pool.ok_to_unbond_other_with(&caller, &target, &delegator)?;
+
+			// Claim the the payout prior to unbonding. Once the user is unbonding their points
+			// no longer exist in the bonded pool and thus they can no longer claim their payouts.
+			// It is not strictly necessary to claim the rewards, but we do it here for UX.
+			Self::do_reward_payout(target.clone(), delegator, &bonded_pool)?;
+
+			// Re-fetch the delegator because they where updated by `do_reward_payout`.
+			let mut delegator =
```

Comment:
Can't do_reward_payout not write a new delegator to storage, and instead mutate it in memory (as &mut Delegator)?

Reply:
Definitley, but for the sake of keeping things simple I wanted to encapsulate all the logic to the call. Keeping this open so I can come back when cleaning and possibly make the changes